# Import the packages

In [2]:
import pandas as pd 
import numpy as np
from sklearn import preprocessing
from sklearn import tree

# Loading Train Dataset

In [5]:
train_data=pd.read_csv('C:/Python/Day-24/Dataset/train.csv')

In [6]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Checking for Null values

In [7]:
train_data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

# Now we are using Label Encoder

In [9]:
label_encoder = preprocessing.LabelEncoder() 
encode_sex=label_encoder.fit_transform(train_data['Sex']) #fit_transform work as coverting text value into numrical value


# Now we are going to use Tree Model so our dependent variable is categerised so we are using Classifier


In [26]:
tree_model=tree.DecisionTreeClassifier()
predicators=pd.DataFrame([encode_sex,train_data['Age'],train_data['Fare']]).T
tree_model.fit(X=predicators,y=train_data["Survived"])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [35]:
with open ('Titanic_surview.dot','w') as f:
    f=tree.export_graphviz(tree_model,feature_names=['Sex','Age','Fare'],out_file=f)

### In above code we created one file as Dtree1.dot & passed with one variable named f & in variable we stored using tree fucntion we epxport graphviz & here we used tree model &  feature's as Sex,Age,Fare & used out_file for stroing the tree data in that location


# here we are going to calcualte Score of our model

In [36]:
tree_model.score(X=predicators,y=train_data['Survived'])


0.9763779527559056


# The model is 97.6% Accurate and Hence can be used to decide which all pasengers survived and which did not


In [33]:
from sklearn.tree import export_text

In [34]:
feature=["Sex","Age","Fare"]
r=export_text(tree_model,feature_names=feature)
print(r)

|--- Sex <= 0.50
|   |--- Fare <= 48.20
|   |   |--- Fare <= 27.82
|   |   |   |--- Fare <= 25.70
|   |   |   |   |--- Fare <= 24.07
|   |   |   |   |   |--- Fare <= 10.48
|   |   |   |   |   |   |--- Fare <= 8.04
|   |   |   |   |   |   |   |--- Fare <= 6.99
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Fare >  6.99
|   |   |   |   |   |   |   |   |--- Fare <= 7.52
|   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |--- Fare >  7.52
|   |   |   |   |   |   |   |   |   |--- Age <= 14.50
|   |   |   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |   |   |--- Age >  14.50
|   |   |   |   |   |   |   |   |   |   |--- Age <= 17.50
|   |   |   |   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |   |   |   |   |   |--- Age >  17.50
|   |   |   |   |   |   |   |   |   |   |   |--- truncated branch of depth 11
|   |   |   |   |   |   |--- Fare >  8.04
|   |   |   |   |   |   |   |--- Age <= 19